### PR TITLE
fix: add backtracking to opt decoding

### PR DESCRIPTION
**Pull request by crusso**
# Description

Fixes decoding of variants under option types, allowing compatible tagged values from wider options, defaulting incompatible tagged values to null.

Brings decoding of options closer to spec, but see child PR #981 for the remaining edge cases.

This PR is marked `do not merge` until #981 is merged into this one.

# How Has This Been Tested?

See PR for additional candid tests.

# Checklist:

- [x] My changes follow the guidelines in [CONTRIBUTING.md](https://github.com/dfinity/agent-js/blob/main/CONTRIBUTING.md).
- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [ ] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

What CHANGELOG?

### CODE REVIEW

File: packages/candid/src/idl.ts [hunk 1]
```diff
@@ -921,8 +921,19 @@ export class OptClass<T> extends ConstructType<[T] | []> {
...
-        return [this._type.decodeValue(b, opt._type)];
+      case 1: {
+        const checkpoint = b.save();
+        try {
+          const v = this._type.decodeValue(b, opt._type)
+          return [v];
+        } catch (e : any) {
+          b.restore(checkpoint);
+          // skip value at wire type (to advance b)
+          const v = opt._type.decodeValue(b, opt._type)
```
Comment (krpeacock) [inline packages/candid/src/idl.ts:932]: Why do we decode the value again if we are always returning none?

Comment (crusso) [reply packages/candid/src/idl.ts:932]: Because we need to skip past the value so the buffer is consumed...

Comment (crusso) [reply packages/candid/src/idl.ts:932]: The motoko code does this a bit differently. It avoids reparsing the candid by parsing to completion and returning a proper value or sentinel value to indicate coercion failure. But that's harder to integrate here without introducing mistakes.

File: packages/candid/src/idl.test.ts [hunk 2]
```diff
@@ -698,3 +698,31 @@ test('should decode matching optional fields if wire type contains additional fi
...
+    '4449444c036c0161016e026b03787f797f7a7f01000100', // Motoko: {a = ?#x} : {a : ?{#x;#y;#z}}
+    'extended variant under opt expected tag',
+  );
+  testDecode(
+    IDL.Record({a: IDL.Opt(IDL.Variant({ x: IDL.Null, y: IDL.Null }))}),
+    {a: []},
+    '4449444c036c0161016e026b03787f797f7a7f01000102', // Motoko: {a = ?#z} : {a : ?{#x;#y;#z}}
+    'extended variant under opt unexpected tag - defaulting',
+  );
+});
```
Comment (krpeacock) [inline packages/candid/src/idl.test.ts:728]: Some basic reformatting so the individual failures will actually be labeled if the tests failed
```suggestion
describe('IDL opt variant decoding', () => {
  it('should handle the first of three variants', () => {
    testDecode(
      Record({ a: Opt(Variant({ x: Null, y: Null, z: Null })) }),
      { a: [{ x: null }] },
      '4449444c036c0161016e026b03787f797f7a7f01000100', // Motoko: {a = ?#x} : {a : ?{#x;#y;#z}},
      'same variant under opt x',
    );
  });
  it('should handle the third of three variants', () => {
    testDecode(
      Record({ a: Opt(Variant({ x: Null, y: Null, z: Null })) }),
      { a: [{ z: null }] },
      '4449444c036c0161016e026b03787f797f7a7f01000102', // Motoko: {a = ?#z} : {a : ?{#x;#y;#z}}
      'same variant under opt z',
    );
  });
  it('should handle the first of two variants', () => {
    testDecode(
      Record({ a: Opt(Variant({ x: Null, y: Null })) }),
      { a: [{ x: null }] },
      '4449444c036c0161016e026b03787f797f7a7f01000100', // Motoko: {a = ?#x} : {a : ?{#x;#y;#z}}
      'extended variant under opt expected tag',
    );
  });
  it('should handle the option when the option is empty', () => {
    testDecode(
      Record({ a: Opt(Variant({ x: Null, y: Null })) }),
      { a: [] },
      '4449444c036c0161016e026b03787f797f7a7f01000102', // Motoko: {a = ?#z} : {a : ?{#x;#y;#z}}
      'extended variant under opt unexpected tag - defaulting',
    );
  });
});
```

File: packages/candid/src/idl.ts [hunk 3]
```diff
@@ -921,8 +921,19 @@ export class OptClass<T> extends ConstructType<[T] | []> {
...
+          // retun none
+          return [];
+        }
+      }
```
Comment (krpeacock) [inline packages/candid/src/idl.ts:936]: More comments for small-brain people like me trying to understand what's happening
```suggestion
      case 1: {
        // Save the current state of the Pipe `b` to allow rollback in case of an error
        const checkpoint = b.save();
        try {
          // Attempt to decode a value using the `_type` of the current instance
          const v = this._type.decodeValue(b, opt._type);
          // If decoding succeeds, return the value wrapped in an array
          return [v];
        } catch (e: any) {
          // If an error occurs during decoding, restore the Pipe `b` to its previous state
          b.restore(checkpoint);
          // Skip the value at the current wire type to advance the Pipe `b` position
          opt._type.decodeValue(b, opt._type);
          // Return an empty array to indicate a `none` value
          return [];
        }
      }
```

File: packages/candid/src/idl.test.ts [hunk 4]
```diff
@@ -698,3 +698,39 @@ test('should decode matching optional fields if wire type contains additional fi
     b: ['123'],
   });
 });
+
+
+describe('IDL opt variant decoding', () => {
```
Comment (jasonz-dfinity) [inline packages/candid/src/idl.test.ts:703]: It looks like the first 3 test cases can be tested for round trip (use `test_` instead of `testDecode`)?

File: packages/candid/src/idl.test.ts [hunk 5]
```diff
@@ -698,3 +698,39 @@ test('should decode matching optional fields if wire type contains additional fi
...
+      { a: [{ x: null }] },
+      '4449444c036c0161016e026b03787f797f7a7f01000100', // Motoko: {a = ?#x} : {a : ?{#x;#y;#z}}
+      'extended variant under opt expected tag',
+    );
+  });
+  it('should handle the option when the option is empty', () => {
+    testDecode(
+      IDL.Record({ a: IDL.Opt(IDL.Variant({ x: IDL.Null, y: IDL.Null })) }),
+      { a: [] },
+      '4449444c036c0161016e026b03787f797f7a7f01000102', // Motoko: {a = ?#z} : {a : ?{#x;#y;#z}}
```
Comment (jasonz-dfinity) [inline packages/candid/src/idl.test.ts:732]: IMHO if you put this case right after the second test case, it would be more easily understood without knowledge of motoko (as they have the same binary format), assuming (most?) humans can't parse candid binary 😄

File: packages/candid/src/idl.test.ts (hunk 4 above)
Comment (crusso) [reply packages/candid/src/idl.test.ts:703]: Tried that but they don't round-trip because the encoding was produced by motoko. There is no unique encoding in Candid so this would actually be too much to expect, I think.

File: packages/candid/src/idl.test.ts (hunk 5 above)
Comment (crusso) [reply packages/candid/src/idl.test.ts:732]: You can't read the binary?

Comment (crusso) [reply packages/candid/src/idl.test.ts:732]: I was organizing these cases differently - parsing value encoded at expected type and parsing values encoded at  subtype of expected type, with expected tag or unexpected tag.

Comment (crusso) [reply packages/candid/src/idl.test.ts:732]: Anyway, I hope you don't mind but I'm closing this PR in favor of the more comprehensive child PR #981 (which has more tests too).

**Comment by crusso** [comment]
Closing in favor of #981 (which incorporates these changes)
